lib/state_change_syncer.go

```diff
@@ -0,0 +1,726 @@
+
+	// This map is used to keep track of the key and value pairs that state syncer has already tracked (and therefore
+	// don't need to be re-emitted to the state change file).
+	// The key is the stringifyed key of the entry, plus the operation type.
+	// The value is the badger entry that was flushed to the db.
+	MempoolKeyValueMap map[string][]byte
+	// This map tracks the keys that were flushed to the mempool in a single flush.
+	// This is used to determine if there are any tracked mempool transactions that have "disappeared" from the current
+	// mempool state. If so, state syncer should completely refresh the mempool.
+	MempoolFlushKeySet map[string]bool
+
+	// Tracks the flush IDs of the last block sync flush and the last mempool flush.
+	// These are not used during hypersync, as many flushes are being processed asynchronously.
+	BlockSyncFlushId uuid.UUID
+	MempoolFlushId   uuid.UUID
+
+	// Mutex to prevent concurrent writes to the state change file.
+	StateSyncerMutex *sync.Mutex
+	EntryCount       uint64
+
+	BlockHeight uint64
+
+	SyncType NodeSyncType
+
+	// During blocksync, we flush all entries by index to the state change file once the sync is complete.
+	// This is used to track whether this procedure has been initiated.
+	BlocksyncCompleteEntriesFlushed bool
+}
+
+// Open a file, create if it doesn't exist.
+func openOrCreateLogFile(fileName string) (*os.File, error) {
+	file, err := os.OpenFile(fileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return nil, err
+	}
+	return file, nil
+}
+
+// NewStateChangeSyncer initializes necessary log files and returns a StateChangeSyncer.
+func NewStateChangeSyncer(stateChangeFilePath string, nodeSyncType NodeSyncType) *StateChangeSyncer {
+	stateChangeIndexFilePath := fmt.Sprintf("%s-index", stateChangeFilePath)
+	stateChangeMemPoolFilePath := fmt.Sprintf("%s-mempool", stateChangeFilePath)
+	stateChangeMempoolIndexFilePath := fmt.Sprintf("%s-mempool-index", stateChangeFilePath)
+	stateChangeFile, err := openOrCreateLogFile(stateChangeFilePath)
+	if err != nil {
+		glog.Fatalf("Error opening stateChangeFile: %v", err)
+	}
+	stateChangeIndexFile, err := openOrCreateLogFile(stateChangeIndexFilePath)
+	if err != nil {
+		glog.Fatalf("Error opening stateChangeIndexFile: %v", err)
+	}
+	stateChangeMempoolFile, err := openOrCreateLogFile(stateChangeMemPoolFilePath)
+	stateChangeFileInfo, err := stateChangeFile.Stat()
+	if err != nil {
+		glog.Fatalf("Error getting stateChangeFileInfo: %v", err)
+	}
+	stateChangeMempoolIndexFile, err := openOrCreateLogFile(stateChangeMempoolIndexFilePath)
+	if err != nil {
+		glog.Fatalf("Error opening stateChangeMempoolIndexFile: %v", err)
+	}
+
+	stateChangeMempoolFileInfo, err := stateChangeMempoolFile.Stat()
+	if err != nil {
+		glog.Fatalf("Error getting stateChangeMempoolFileInfo: %v", err)
+	}
+
+	// Check if the state change file is empty. If not, BlocksyncCompleteEntriesFlushed should be true.
```

Comment:
I think there is an edge case here where the sync was interrupted in the middle of flushing entries to the log file. If true, just add a comment explaining that.

Reply:
Also if that ever happens it would be nice if you could detect it and kill the node so that it doesn't run with a corrupted state. I'd add that as a TODO underneath.

Reply:
Having read more of the code now, it seems like you're good as long as there's a StateSyncerFlushed thing somewhere.